PetrolCar.cpp

```diff
@@ -1,16 +1,39 @@
 #include "PetrolCar.hpp"
-#include <iostream>
- 
-PetrolCar::PetrolCar(PetrolEngine* engine)
-    : engine_(engine)
+
+PetrolCar::PetrolCar(std::unique_ptr<PetrolEngine> engine)
+    : engine_(std::move(engine))
 {
     std::cout << __FUNCTION__ << std::endl;
 }
 
-PetrolCar::~PetrolCar()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::refuel()        { std::cout << __FUNCTION__ << std::endl; }
+PetrolCar::~PetrolCar() { std::cout << __FUNCTION__ << std::endl; }
+void PetrolCar::refuel() { std::cout << __FUNCTION__ << std::endl; }
+
+void PetrolCar::changeEngine(std::unique_ptr<PetrolEngine> engine)
+{
+    engine_ = std::move(engine);
+    std::cout << __FUNCTION__ << std::endl;
+}
+
+void PetrolCar::changeEngine(std::unique_ptr<ElectricEngine> engine)
+{
+    throw std::invalid_argument("This is not electric Car");
+}
+
+void PetrolCar::refill()
+{
+    refuel();
+    std::cout << __FUNCTION__ << std::endl;
+}
 
+void PetrolCar::changeGear(int gear)
+{
+    try
+    {
+        engine_->changeGear(gear);
+    }
+    catch(const InvalidGear& e)
+    {
+        throw e;
+    }
```

Comment:
This try/catch doesn't make sense. Just remove it.
You rethrow the same exception and don't do anything more here.